wagtail/core/models.py

```diff
@@ -1830,8 +1976,45 @@ def is_latest_revision(self):
         latest_revision = PageRevision.objects.filter(page_id=self.page_id).order_by('-created_at', '-id').first()
         return (latest_revision == self)
 
-    def publish(self):
+    def publish(self, user=None, changed=True, log_action=True, previous_revision=None):
+        """
+        Publishes or schedules revision for publishing.
+
+        :param user: the publishing user
+        :param changed: indicated whether content has changed
+        :param log_action:
+            flag for the logging action. Pass False to skip logging. Cannot pass an action string as the method
+            performs several actions: "publish", "revert" (and publish the reverted revision),
+            "schedule publishing with a live revision", "schedule revision reversal publishing, with a live revision",
+            "schedule publishing", "schedule revision reversal publishing"
+        :param previous_revision: indicates a revision reversal. Should be set to the previous revision instance
+        """
         page = self.as_page_object()
+
+        def log_scheduling_action(revision, previous_revision=None, user=None, changed=changed):
```

Comment:
On scheduling, I think we might want to set the `user` attribute here: https://github.com/wagtail/wagtail/blob/master/wagtail/core/management/commands/publish_scheduled_pages.py#L108

To match the user in the revision?

Also, should we have separate action types for publishing from a schedule and a page expiring? (we display expiry differently in the UI to unpublish)

Reply:
Partially solved in a79b2fb